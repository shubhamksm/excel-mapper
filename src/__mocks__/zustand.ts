import { StoreState, useBoundStore } from "@/store/useBoundStore";

import type * as ZustandExportedTypes from "zustand";
export * from "zustand";

// required actual to inherit all functionality in tests
jest.requireActual<typeof ZustandExportedTypes>("zustand");

// Turn useBoundStore to mock
jest.mock("../store/useBoundStore", () => ({
  useBoundStore: jest.fn(),
}));

// Using jest.mock will allow retain the types to useStoreMock
const useStoreMock = jest.mocked(useBoundStore);

// Individual tests can use this to override
export const mockUseStore = (overrides: Partial<StoreState> = {}) => {
  useStoreMock.mockImplementation((getterFn) => {
    return getterFn({
      // we include the store's actual values by default
      // this allows the mocked store to have complete functionality,
      // with "granular" mocks defined as specified by tests
      ...jest.requireActual("../store/useBoundStore").useBoundStore(),
      ...overrides,
    });
  });
};

// this will set the default mock for the store on a per-test basis
// Note: setting this mock per test is a little heavy-handed, alternatively
// you can use beforeAll to set the default mock once per test suite
beforeEach(() => {
  mockUseStore();
});
