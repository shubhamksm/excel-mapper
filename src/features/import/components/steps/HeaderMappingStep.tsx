import { useMemo } from "react";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";
import { useBoundStore } from "@/features/import/store/useBoundStore";
import { extractHeaders, mapRowWithHeaders } from "@/utils";
import { REQUIRED_TEMPLATE_COLUMNS, TEMPLATE_COLUMNS } from "@/constants";
import type { Template_Columns } from "@/types";
import { useShallow } from "zustand/react/shallow";
import { ModalFooter } from "../ModalFooter";
import { ToggleGroup, ToggleGroupItem } from "@/components/ui/toggle-group";

export const HeaderMappingStep = () => {
  const [mappedHeaders, setMappedHeaders, parsedFile, setTitleMappedData] =
    useBoundStore(
      useShallow((state) => [
        state.mappedHeaders,
        state.setMappedHeaders,
        state.parsedFile,
        state.setTitleMappedData,
      ])
    );

  const isHeaderMappingNextButtonDisabled = useMemo(() => {
    return (
      Object.values(mappedHeaders).filter(({ column }) =>
        REQUIRED_TEMPLATE_COLUMNS.includes(column)
      ).length < REQUIRED_TEMPLATE_COLUMNS.length
    );
  }, [mappedHeaders]);

  const headers = useMemo(() => {
    if (parsedFile) {
      return extractHeaders(parsedFile[0]);
    }
    return [];
  }, [parsedFile]);

  const handleChangeColumn = (header: string, value: Template_Columns) => {
    setMappedHeaders({
      ...mappedHeaders,
      [header]: {
        column: value,
        ...(value === "amount" ? { debitOrCredit: "both" } : {}),
      },
    });
  };

  const handleChangeDebitOrCredit = (
    header: string,
    value: "debit" | "credit" | "both"
  ) => {
    setMappedHeaders({
      ...mappedHeaders,
      [header]: { ...mappedHeaders[header], debitOrCredit: value },
    });
  };

  const handleNext = () => {
    if (!parsedFile) {
      throw new Error("No file uploaded");
    }
    const mappedData = mapRowWithHeaders(parsedFile, mappedHeaders);
    setTitleMappedData(mappedData);
  };

  return (
    <>
      <div className="space-y-4 h-full flex-1 overflow-y-auto w-full">
        <p className="text-sm text-muted-foreground text-center">
          Map your file headers to the corresponding fields in our system.
        </p>
        {headers.map((header) => (
          <div key={header} className="flex items-center justify-between gap-4">
            <span className="text-sm font-medium">{header}</span>
            <div className="flex items-center gap-2">
              {mappedHeaders[header]?.column === "amount" && (
                <ToggleGroup
                  type="single"
                  value={mappedHeaders[header]?.debitOrCredit}
                  onValueChange={(value) =>
                    handleChangeDebitOrCredit(
                      header,
                      value as "debit" | "credit" | "both"
                    )
                  }
                  className="scale-75 origin-right"
                >
                  <ToggleGroupItem
                    value="debit"
                    className="text-red-500 data-[state=on]:text-red-500"
                  >
                    <span className="text-2xl"> - </span>
                  </ToggleGroupItem>
                  <ToggleGroupItem
                    value="credit"
                    className="text-green-500 data-[state=on]:text-green-500"
                  >
                    <span className="text-2xl"> + </span>
                  </ToggleGroupItem>
                  <ToggleGroupItem
                    value="both"
                    className="data-[state=on]:bg-gradient-to-t from-red-500 to-green-500 data-[state=on]:text-white"
                  >
                    <span className="text-2xl"> Â± </span>
                  </ToggleGroupItem>
                </ToggleGroup>
              )}
              <Select
                onValueChange={(value) =>
                  handleChangeColumn(header, value as Template_Columns)
                }
              >
                <SelectTrigger className="w-[180px]">
                  <SelectValue placeholder="Select field" />
                </SelectTrigger>
                <SelectContent>
                  {TEMPLATE_COLUMNS.map((column) => (
                    <SelectItem key={column} value={column}>
                      {column.charAt(0).toUpperCase() + column.slice(1)}
                    </SelectItem>
                  ))}
                </SelectContent>
              </Select>
            </div>
          </div>
        ))}
      </div>
      <ModalFooter
        isNextDisabled={isHeaderMappingNextButtonDisabled}
        onNext={handleNext}
      />
    </>
  );
};
