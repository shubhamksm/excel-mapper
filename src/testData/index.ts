export const parsedFile1 = [
  {
    Type: "TOPUP",
    Product: "Current",
    "Started Date": "2022-10-04 12:40:58",
    "Completed Date": "2022-10-04 12:42:41",
    Description: "Card Top-Up",
    Amount: "83.99",
    Fee: "0.00",
    Currency: "NOK",
    State: "COMPLETED",
    Balance: "83.99",
  },
  {
    Type: "CARD_PAYMENT",
    Product: "Current",
    "Started Date": "2022-10-06 13:06:09",
    "Completed Date": "2022-10-08 16:27:29",
    Description: "REMA 1000 LAKKEGATA",
    Amount: "-52.52",
    Fee: "0.00",
    Currency: "NOK",
    State: "COMPLETED",
    Balance: "1947.48",
  },
  {
    Type: "TRANSFER",
    Product: "Savings",
    "Started Date": "2022-10-09 09:15:00",
    "Completed Date": "2022-10-09 09:16:30",
    Description: "Transfer to Savings",
    Amount: "-500.00",
    Fee: "0.00",
    Currency: "NOK",
    State: "COMPLETED",
    Balance: "1447.48",
  },
  {
    Type: "ATM_WITHDRAWAL",
    Product: "Current",
    "Started Date": "2022-10-10 14:20:00",
    "Completed Date": "2022-10-10 14:21:15",
    Description: "ATM Withdrawal",
    Amount: "-200.00",
    Fee: "2.00",
    Currency: "NOK",
    State: "COMPLETED",
    Balance: "1245.48",
  },
];
