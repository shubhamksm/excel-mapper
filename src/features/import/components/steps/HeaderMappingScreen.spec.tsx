import { mockUseStore } from "@/__mocks__/zustand";
import { parsedFile1 } from "@/testData";
import { render } from "@testing-library/react";
import { HeaderMappingStep } from "./HeaderMappingStep";

describe("Header Mapping Screen", () => {
  const parsedFile = parsedFile1;
  const changeCurrentScreen = jest.fn();
  const mockState = { parsedFile, changeCurrentScreen };

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseStore(mockState);
  });

  it("Should match snapshots", () => {
    const { container } = render(<HeaderMappingStep />);

    expect(container).toMatchSnapshot();
  });
});
