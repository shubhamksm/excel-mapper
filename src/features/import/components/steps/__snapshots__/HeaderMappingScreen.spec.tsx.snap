// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Mapping Screen Should match snapshots 1`] = `
<div>
  <div
    class="space-y-4 h-full flex-1 overflow-y-auto w-full"
  >
    <p
      class="text-sm text-muted-foreground text-center"
    >
      Map your file headers to the corresponding fields in our system.
    </p>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Type
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r0:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Product
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r1:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Started Date
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r2:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Completed Date
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r3:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Description
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r4:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Amount
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r5:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Fee
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r6:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Currency
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r7:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        State
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r8:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex items-center justify-between gap-4"
    >
      <span
        class="text-sm font-medium"
      >
        Balance
      </span>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r9:"
          aria-expanded="false"
          class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Select field
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="flex justify-between mt-4"
  >
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
      disabled=""
    >
      Previous
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;
